src/librustc_mir/const_eval/mod.rs

```diff
@@ -67,3 +69,32 @@ pub(crate) fn destructure_const<'tcx>(
 
     mir::DestructuredConst { variant, fields }
 }
+
+pub(crate) fn deref_const<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    val: &'tcx ty::Const<'tcx>,
+) -> &'tcx ty::Const<'tcx> {
+    trace!("deref_const: {:?}", val);
+    let ecx = mk_eval_cx(tcx, DUMMY_SP, param_env, false);
+    let op = ecx.eval_const_to_op(val, None).unwrap();
+    let mplace = ecx.deref_operand(op).unwrap();
+
+    let ty = match mplace.meta {
+        MemPlaceMeta::None => mplace.layout.ty,
+        MemPlaceMeta::Poison => bug!("poison metadata in `deref_const`: {:#?}", mplace),
+        // In case of unsized types, figure out the real type behind.
+        MemPlaceMeta::Meta(scalar) => match mplace.layout.ty.kind {
+            ty::Dynamic(..) => ecx.read_drop_type_from_vtable(scalar).unwrap().1,
```

Comment:
I really don't want us to be doing this, it's magically extracting a *lifetime-erased* version of the original type from a vtable, something that is otherwise impossible.

Note that the lifetime-erased aspect might interact poorly with borrow-checking, but I'm not sure how much of a problem that is in practice. (And I guess `destructure_const` has the same problem, ugh - maybe we need to get the field types from original type itself, and not rely on the `TyAndLayout`?)

IMO `deref_const` should produce an unsized `ty::Const` instead of trying to divine a sized type.

If we don't want `ty::Const` to be able to represent unsized values, we can offer a different interface (at least for slices).

Reply:
yea, I'll make it bug out. I mainly added it because I could.